### PR TITLE
Feature - Database implementation

### LINKED ISSUE #363: Rework: Database rework and planning

### PR DESCRIPTION
**Pull request by v-stamenova**
# Description

This branch introduces the initial design of the database. The ERD of the database can be seen below:
conference-diagram drawio (4)
Introduces as well helper methods in models `User` and `Presentation` in order to hide the `UserPresentation` model.

closes #363 #364 

## Type of change
- [X] New feature (non-breaking change which adds functionality)
- [X] This change requires a documentation update

## What needs to be tested

- [ ] After migrating and seeding (`sail artisan migrate:fresh --seed`), might be a good idea to play around with `sail tinker` and check if the relationship established between the models are actually established correctly

# Checklist:

- [X] I have performed a self-review of my code
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by IGORnvk — COMMENTED** [review]
The relationships work perfectly! You did a great job. There are just some minor issues that I found that do not align with the ERD, but other than that, everything looks good to me.

**Comment by IGORnvk** [comment]
image
Also, in `notifications` table on the ERD, `notifiable_type` is mentioned twice, but has different data types. One of them should probably be renamed to `notifiable_id`. In code it is correct, just the ERD.

File: database/migrations/2024_03_25_161038_create_companies_table.php [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        Schema::create('companies', function (Blueprint $table) {
+            $table->id();
+
+            $table->string('name');
+            $table->string('website');
+            $table->text('description');
+            $table->boolean('is_approved')->default(false);
+            $table->unsignedBigInteger('sponsorship_id')->nullable();
+            $table->boolean('is_sponsorship_approved')->nullable();
+            $table->unsignedBigInteger('booth_id')->nullable();
```
[the remark this answers is not in the document]
Comment (v-stamenova) [reply database/migrations/2024_03_25_161038_create_companies_table.php:23]: Good catch, initially I thought of adding the foreign key in the relationship between `Booth` and `Company` in the `Company` model but decided against it.

**Comment by v-stamenova** [comment]
> image Also, in `notifications` table on the ERD, `notifiable_type` is mentioned twice, but has different data types. One of them should probably be renamed to `notifiable_id`. In code it is correct, just the ERD.

Uploaded the new version of the ERD, thanks!

**Review by IGORnvk — APPROVED** [review]

File: app/Models/User.php [hunk 2]
```diff
@@ -17,6 +22,111 @@ class User extends Authenticatable
...
+            'user_id' => $this->id,
+            'presentation_id' => $presentation->id,
+            'role' => $role
+        ]);
+
+        return true;
+    }
+
+    /**
+     * Dis-enrol participant from a presentation
```
Comment (TimKardol) [inline app/Models/User.php:86]:
```suggestion
     * Unenroll participant from a presentation
```
Nitpicky, but this is the correct grammar.

File: database/factories/BoothFactory.php
Comment (TimKardol) [inline database/factories/BoothFactory.php:1]: I am guessing `width` and `length` are in meters. Maybe specify that in the comment. Also, 10x10m is a 100 square meters. That's a lot. Maybe change the min and max value when we have specific sizes.

File: database/migrations/2024_03_26_094332_create_presentations_table.php [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+            $table->unsignedBigInteger('room_id')->nullable();
+            $table->unsignedBigInteger('difficulty_id')->nullable();
+            $table->unsignedBigInteger('company_id')->nullable();
+
+            $table->foreign('timeslot_id')->references('id')->on('timeslots')
+                ->cascadeOnUpdate()->nullOnDelete();
+            $table->foreign('room_id')->references('id')->on('rooms')
+                ->cascadeOnUpdate()->nullOnDelete();
+            $table->foreign('difficulty_id')->references('id')->on('difficulties')
+                ->cascadeOnUpdate()->nullOnDelete();
```
Comment (dwaard) [inline database/migrations/2024_03_26_094332_create_presentations_table.php:37]: change to restrictOnDelete()

File: database/migrations/2024_03_26_094332_create_presentations_table.php [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+            $table->foreign('company_id')->references('id')->on('companies')
+                ->cascadeOnUpdate()->cascadeOnDelete();
```
Comment (dwaard) [inline database/migrations/2024_03_26_094332_create_presentations_table.php:39]: Check the business if presentation must be deleted when a company is deleted

**Review by dwaard — APPROVED** [review]

**Pull request merged**